src/Features/Core/Portable/SolutionCrawler/SolutionCrawlerProgressReporter.cs

```diff
@@ -19,109 +16,69 @@ internal partial class SolutionCrawlerRegistrationService : ISolutionCrawlerRegi
+            }
 
-            // this is to reduce number of times we report progress on same file. this is purely for perf
-            private string _lastReportedFilePath;
+            public event EventHandler<ProgressData> ProgressChanged;
 
-            private int _count;
+            public bool InProgress => _progressStartCount > 0;
 
-            public SolutionCrawlerProgressReporter(IAsynchronousOperationListener listener)
-            {
-                _listener = listener;
-                _eventQueue = new SimpleTaskQueue(TaskScheduler.Default);
-                _eventMap = new EventMap();
+            public void Start() => ChangeProgressStatus(ref _progressStartCount, ProgressStatus.Started);
+            public void Stop() => ChangeProgressStatus(ref _progressStartCount, ProgressStatus.Stoped);
+
+            private void Evaluate() => ChangeProgressStatus(ref _progressEvaluateCount, ProgressStatus.Evaluating);
+            private void Pause() => ChangeProgressStatus(ref _progressEvaluateCount, ProgressStatus.Paused);
 
-                _count = 0;
-            }
 
-            public bool InProgress
+            public void UpdatePendingItemCount(int pendingItemCount)
             {
-                get
+                if (_progressStartCount > 0)
                 {
-                    return _count > 0;
+                    var progressData = new ProgressData(ProgressStatus.PendingItemUpdated, pendingItemCount);
+                    OnProgressChanged(progressData);
                 }
             }
 
-            public event EventHandler<ProgressData> ProgressChanged
+            public IDisposable Evaluating()
             {
-                add
-                {
-                    _eventMap.AddEventHandler(nameof(ProgressChanged), value);
-                }
-
-                remove
-                {
-                    _eventMap.RemoveEventHandler(nameof(ProgressChanged), value);
-                }
+                return new ProgressStatusRAII(this);
             }
 
-            public Task Start()
+            private void ChangeProgressStatus(ref int referenceCount, ProgressStatus status)
             {
-                if (Interlocked.Increment(ref _count) == 1)
+                var start = (status == ProgressStatus.Started || status == ProgressStatus.Evaluating);
+                if (start ? (Interlocked.Increment(ref referenceCount) == 1) : (Interlocked.Decrement(ref referenceCount) == 0))
                 {
-                    _lastReportedFilePath = null;
-
-                    var asyncToken = _listener.BeginAsyncOperation("ProgressReportStart");
-                    var progressData = new ProgressData(ProgressStatus.Started, filePathOpt: null);
-                    return RaiseEvent(nameof(ProgressChanged), progressData).CompletesAsyncOperation(asyncToken);
+                    var progressData = new ProgressData(status, pendingItemCount: null);
+                    OnProgressChanged(progressData);
```

Comment:
what's the threading guarantee for this event.  For example, imagine we have the following two things happen concurrently:

1. a thread comes in and causes us to switch to the 'evaluating state'.  then, it doesn't get to OnProgressChanged.
2. a thread comes in and causes us to switch to teh 'paused state'.  it then issues OnProgressChanged(paused).
3. first thread wakes up and issues OnProgressChanged(evaluating)

However,  in this case, the the listenener will get the events out of order which could lead to the UI being stuck in teh 'evaluating' state.

Reply:
Should we serialize (not necessarily on hte UI thread) this stuff to ensure the UI is always in a consistent state wrt our internal state?

Reply:
I had that and got rid of it. those race hardly happen and when it happens it doesn't matter much. in the end, UI goes away/cleaned up.